Proofs/Experiments/MemoryAliasing.lean

```diff
@@ -8,22 +8,586 @@ The goal is to eliminate the sorry, and to simplify the proof to a tactic invoca
+
+@[simp]
+theorem BitVec.getLsb_extractLsB (val : BitVec w₁) :
+    ((BitVec.extractLsB val n).getLsb i) =
+    (decide (i ≤ 7) && val.getLsb (n * 8 + i)) := by
+  simp [extractLsB]
+  rw [Nat.succ_mul]
+  simp only [Nat.add_one_sub_one]
+  simp [Nat.add_sub_cancel_left]
+
+/-- Writing zero bytes does not change memory. -/
+theorem write_mem_bytes'_zero : write_mem_bytes' 0 addr val s = s := rfl
+
+/-- Writiing (n + 1) bytes can be described as writing `n` bytes, and then recursing to write the rest. -/
+theorem write_mem_bytes'_succ :
+    write_mem_bytes' (n + 1) addr val s =
+    let byte := BitVec.extractLsb 7 0 val
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n * 8) (val >>> 8) -- TODO: rewrite this as 'truncate'.
+    write_mem_bytes' n (addr + 1#64) val_rest s := rfl
+
+def write_mem_bytes'₂ (n : Nat) (addr : BitVec 64) (val : BitVec (n * 8)) (s : Memory) : Memory :=
+  match n with
+  | 0 => s
+  | n' + 1 =>
+    let byte := BitVec.extractLsB val 0 -- rewrite in terms of extractLsB
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n' * 8) (val >>> 8)
+    write_mem_bytes' n' (addr + 1#64) val_rest s
+
+theorem write_mem_bytes'_eq_write_mem_bytes'₂ :
+    write_mem_bytes' n addr val s = write_mem_bytes'₂ n addr val s := by
+  induction n generalizing s addr
+  case zero => simp [write_mem_bytes', write_mem_bytes'₂]
+  case succ n ih =>
+    simp [write_mem_bytes', write_mem_bytes'₂]
+    simp [ih, BitVec.extractLsB]
+
+theorem BitVec.neq_of_lt {x y : BitVec w₁} (h : x < y) : x ≠ y := by
+  intros h'
+  subst h'
+  simp [BitVec.lt_def] at h
+
+theorem BitVec.neq_of_gt {x y : BitVec w₁} (h : x > y) : x ≠ y := by
+  intros h'
+  subst h'
+  simp [BitVec.lt_def] at h
+
+
+/-- adding bitvectors when there is no overflow. -/
+theorem BitVec.toNat_add_eq_toNat_add_toNat {x y : BitVec w₁} (h : x.toNat + y.toNat < 2^w₁) : (x + y).toNat = x.toNat + y.toNat := by
+  rw [BitVec.toNat_add, Nat.mod_eq_of_lt h]
+
+theorem write_mem_bytes'_eq_of_lt {ix base : BitVec 64} (hix : ix.toNat < base.toNat) (hnowrap : base.toNat + n < 2^64) :
+    write_mem_bytes' n base data mem ix = mem ix := by
+  induction n generalizing base mem ix
+  case zero => simp [write_mem_bytes']
+  case succ n ih =>
+    simp [write_mem_bytes']
+    rw [ih]
+    · apply write_mem'_of_neq (BitVec.neq_of_lt hix)
+    · rw [BitVec.toNat_add_eq_toNat_add_toNat (by simp; omega)]
+      omega
+    · rw [BitVec.toNat_add_eq_toNat_add_toNat (by simp; omega)]
+      simp; omega
+
+-- bollu: The use of the identifier `w` in a non-scoped namespace brings me great pain, because
+-- I can't name my widths `w` anymore.
+theorem BitVec.gt_def {x y : BitVec w₁} : (x > y) = (x.toNat > y.toNat) := rfl
+
+-- bollu: The use of the identifier `w` in a non-scoped namespace brings me great pain, because
+-- I can't name my widths `w` anymore.
+theorem BitVec.ge_def {x y : BitVec w₁} : (x ≥ y) = (x.toNat ≥ y.toNat) := rfl
```

Comment:
I think this is worth upstreaming. In our code, we use just notation to keep the core calculus small:

```
notation:50 x " ≤ᵤ " y => BitVec.ule x y
notation:50 x " <ᵤ " y => BitVec.ult x y
notation:50 x " ≥ᵤ " y => BitVec.ule y x
notation:50 x " >ᵤ " y => BitVec.ult y x
```

Reply:
I don't think the notation helps here: this theorem is really an equation lemma for `ge`. It would be useful to upstream, indeed.